patchAPI/example/main.cpp

```diff
@@ -0,0 +1,115 @@
+// DynInst
+#include "BPatch.h"
+#include "BPatch_binaryEdit.h"
+#include "BPatch_image.h"
+#include "BPatch_function.h"
+#include "BPatch_object.h"
+#include "BPatch_point.h"
+#include "BPatch_Vector.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <iterator>
+#include <string>
+#include <vector>
+#include <iterator>
+
+// patchAPI
+#include "PatchMgr.h"
+
+using namespace std;
+using namespace Dyninst;
+using namespace PatchAPI;
+
+
+// a simple example, just insert a bunch of no ops
+class NoopSnippet : public Snippet {
+
+public:
+  bool generate(Point *pt, Buffer &buffer){
+    uint8_t byte = 0x90;
+    cout << "inserting a no op @" << pt << endl;
+    for(int i = 0; i < 10; i++){
+      buffer.push_back(byte);
+    }
+    return true;
+  }
+
+};
+
+
+
+int main(int argc, const char *argv[]) {
+
+  if(argc != 3){
+    cerr << "Usage:\n\t" << argv[0] << " <input binary> <output binary path>" << endl;
+    return 1;
+  }
+
+  const char* input_binary = argv[1];
+  const char* output_binary = argv[2];
+
+  BPatch bpatch;
+
+  BPatch_binaryEdit* app = bpatch.openBinary(input_binary, false);
+
+  if(app == NULL){
+    return 0;
+  }
+
+  cout << "app OK" << endl;
+
+  BPatch_image* image = app->getImage();
+
+  cout << "image OK" << endl;
+
+  vector<BPatch_object*> objects;
+
+  image->getObjects(objects);
+
+  cout << "objects: " << objects.size() << endl;
+
+  PatchMgrPtr patchMgr = PatchAPI::convert(image);
+
+  BPatch_object* batchObj = objects[0];
+
+  // Not mentioned in A.2 of https://dyninst.org/sites/default/files/manuals/dyninst/patchAPI.pdf
+  // But found in the header file:  "BPatch_object.h"
+  PatchObject* binobj = PatchAPI::convert(batchObj);
+
+  Patcher patcher(patchMgr);
+
+  NoopSnippet::Ptr snippet = NoopSnippet::create(new NoopSnippet);
+
+  vector<PatchFunction*> functions;
+
+  binobj->funcs(back_inserter(functions));
+
+
+  for(vector<PatchFunction*>::iterator funIter = functions.begin(); funIter != functions.end(); funIter++){
+    PatchFunction *fun = *funIter;
+
+    vector<Point*> f_entryPoints;
```

Comment:
Destruction of this vector leaks all of its containing objects because they are pointers.

Reply:
These objects pointed by the pointers are constructed inside Dynins  and the objects are stored in a map. These objects will be destructed when Dyninst destructs the map. Users should not explicitly destruct these objects.

Reply:
I did valgrind it, but got so much noise that I left it at that.
```
==1932== LEAK SUMMARY:
==1932==    definitely lost: 139,464 bytes in 449 blocks
==1932==    indirectly lost: 218,810 bytes in 2,894 blocks
==1932==      possibly lost: 110,365 bytes in 1,721 blocks
==1932==    still reachable: 1,284,088 bytes in 16,184 blocks
==1932==                       of which reachable via heuristic:
==1932==                         newarray           : 528 bytes in 1 blocks
==1932==                         multipleinheritance: 239,304 bytes in 767 blocks
==1932==         suppressed: 0 bytes in 0 blocks
==1932== Reachable blocks (those to which a pointer was found) are not shown.
==1932== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==1932==
==1932== For counts of detected and suppressed errors, rerun with: -v
==1932== ERROR SUMMARY: 62 errors from 62 contexts (suppressed: 0 from 0)

```